README
WWW::VERNet(3)        User Contributed Perl Documentation       WWW::VERNet(3)



   NNAAMMEE
       WWW::VERNet - Screen-scraping API for VERNet service information.

   SSYYNNOOPPSSIISS
       This module provides a rudimentary screen-scraping "API" for VERNet
       service information.

       You may use this module to access information about your VERNet
       services including live data from end points.

               use WWW::VERNet;

               # Create a new WWW::VERNet object using our customer credentials.

               my $v = WWW::VERNet->new(
                               username => 'email@myorganisation.org',
                               password => 'My%3cretP@ssw0rd!'
                       ) or die "Unable to create new WWW::VERNet object: $!\n";

               # For each VERNet service, print out the service ID and description,
               # and for service endpoint, print out the endpoint ID, link and transmit
               # status, and receive power _if_ available.

               for my $service ( $v->services ) {
                       if ( $service->has_pop_data ) {
                               print "Service ". $service->id ." (". $service->description .")\n";

                               for my $endpoint ( $service->live_data->endpoints ) {

                                       print " - Endpoint ". $endpoint->eid
                                               ."\n\tLink OK: ".       $endpoint->link_ok
                                               ."\n\tReceive power: ". $endpoint->receive_power
                                               ."\n\tTransmitting: ".  $endpoint->pop_transmitting
                                               ."\n\n"
                               }

                               print "\n\n"
                       }
               }

               # e.g. print something like:
               #
               # Service MYORG01234567M (CBD Campus to Remote Campus)
               # - Endpoint A001
               #       Link OK: Yes
               #       Receive power: -5.67 dBm
               #       Transmitting: -5.67 dBm
               #
               # - Endpoint A002
               #       Link OK: Yes
               #       Receive power: -8.096 dBm
               #       Transmitting: -8.096 dBm
               #
               # ...


               # If we know our service and endpoint IDs, we can get direct access
               # to the live data we are interested in.

               print "MYORG01234567M endpoint A001 Rx power is "
                       . $v->service( 'MYORG01234567M' )->live_data->endpoint( 'A001' )->receive_power;

               create_service_alert( 'VERNet service MYORG01234567M link state degraded' )
                       unless ( $v->service( 'MYORG01234567M' )->live_data->endpoint( 'A001' )->link_ok );

   MMEETTHHOODDSS
       _n_e_w _( _%_a_r_g_s _)

       Constructor - accepts the following parameters:

       username
           Mandatory - the username of the VERNet customer account.

       password
           Mandatory - the password of the VERNet customer account.

       timeout
           Optional - the timeout period (in seconds) to wait for HTTP
           requests to be service.  If not provided this will default to a
           value of 180s.

       _s_e_r_v_i_c_e_s _(_)

       Returns an array of WWW::VERNet::Service objects representing the
       provisioned VERNet services for your account.

       _s_e_r_v_i_c_e _( _$_s_e_r_v_i_c_e___i_d _)

       Returns a WWW::VERNet::Service object representing the service
       identified by the $$sseerrvviiccee__iidd parameter.

   AAUUTTHHOORR
       Luke Poskitt, "<ltp at cpan.org>"

   BBUUGGSS
       Please report any bugs or feature requests to "bug-www-vernet at
       rt.cpan.org", or through the web interface at
       <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-VERNet>.  I will be
       notified, and then you'll automatically be notified of progress on your
       bug as I make changes.

   SSEEEE AALLSSOO
       WWW::VERNet::Service WWW::VERNet::Service::LiveData
       WWW::VERNet::Service::LiveData::EndPoint

   SSUUPPPPOORRTT
       PPlleeaassee nnoottee that this is nnoott an authorised or official API and that it
       is nnoott endorsed or supported by VERNet.  This module is a convenience
       provided by community support only.

       You can find documentation for this module with the perldoc command.

           perldoc WWW::VERNet

       You can also look for information at:

       ·   RT: CPAN's request tracker (report bugs here)

           <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-VERNet>

       ·   AnnoCPAN: Annotated CPAN documentation

           <http://annocpan.org/dist/WWW-VERNet>

       ·   CPAN Ratings

           <http://cpanratings.perl.org/d/WWW-VERNet>

       ·   Search CPAN

           <http://search.cpan.org/dist/WWW-VERNet/>

   AACCKKNNOOWWLLEEDDGGEEMMEENNTTSS
   LLIICCEENNSSEE AANNDD CCOOPPYYRRIIGGHHTT
       Copyright 2017 Luke Poskitt.

       This program is free software; you can redistribute it and/or modify it
       under the terms of the the Artistic License (2.0). You may obtain a
       copy of the full license at:

       <http://www.perlfoundation.org/artistic_license_2_0>

       Any use, modification, and distribution of the Standard or Modified
       Versions is governed by this Artistic License. By using, modifying or
       distributing the Package, you accept this license. Do not use, modify,
       or distribute the Package, if you do not accept this license.

       If your Modified Version has been derived from a Modified Version made
       by someone other than you, you are nevertheless required to ensure that
       your Modified Version complies with the requirements of this license.

       This license does not grant you the right to use any trademark, service
       mark, tradename, or logo of the Copyright Holder.

       This license includes the non-exclusive, worldwide, free-of-charge
       patent license to make, have made, use, offer to sell, sell, import and
       otherwise transfer the Package with respect to any patent claims
       licensable by the Copyright Holder that are necessarily infringed by
       the Package. If you institute patent litigation (including a cross-
       claim or counterclaim) against any party alleging that the Package
       constitutes direct or contributory patent infringement, then this
       Artistic License to you shall terminate on the date that such
       litigation is filed.

       Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
       AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
       THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
       PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
       YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
       CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
       CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
       EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



perl v5.18.2                      2017-09-25                    WWW::VERNet(3)


Changes
Revision history for WWW-VERNet

0.01    * Mon Sep 25 2017 ltp@cpan.org 0.01
        - Initial release.



lib/WWW/VERNet/Service.pm
package WWW::VERNet::Service;

use strict;
use warnings;

use Scalar::Util qw( weaken );
use WWW::VERNet::Service::LiveData;

our @ATTR = qw( sid id description ps_link has_ps has_pop_data type );

{
	no strict 'refs';

	for my $attr ( @ATTR ) {
		*{ __PACKAGE__ ."::$attr" } = sub {
			my $self = shift;
			return $self->{ $attr }
		}
	}
}

sub new {
	my ( $class, $t, $s, $v ) = @_;

	my $self = bless {}, $class;
	$self->__init( $t, $s, $v );

	return $self
}

sub __init {
	my ( $self, $t, $s, $v ) = @_;

	weaken ( $self->{ __v }	= $v );
	$self->{ type }		= $t;
	$self->{ description }	= $s->as_trimmed_text;
	$self->{ description }	=~ s/^\s*\S+\s*//;

	( $self->{ sid }, $self->{ ps_link } ) 
				= $s->look_down( _tag => 'a' );

	$self->{ id }		= $self->{ sid }->as_trimmed_text;

	( $self->{ sid }	= $self->{ sid }->attr( 'href' ) ) 
				=~ s/^.*sid=//;
	
	if ( defined $self->{ ps_link } ) {
		$self->{ ps_link }	= $WWW::VERNet::__DOMAIN . $self->{ ps_link }->attr( 'href' );
		$self->{ has_ps }	= 1
	}

	if ( $s->look_down( _tag => 'img', alt => qr/Live data .* is available/ ) ) {
		$self->{ has_pop_data }	= 1
	}
}

sub live_data {
	my $self = shift;
	
	return $self->__live_data;
}

sub __live_data {
	my $self = shift;

	my $url = "https://$WWW::VERNet::__DOMAIN/viewservice.php?sid=$self->{ sid }";

	$self->{ __v }->{ __m }->get( $url );
	
        if ( $self->{ __v }->{ __m }->success
                and $self->{ __v }->{ __m }->title
                and $self->{ __v }->{ __m }->title =~ /Service Details/i
        ) {
                my $l = WWW::VERNet::Service::LiveData->new(
			$self->{ __t } = HTML::TreeBuilder->new_from_content( $self->{ __v }->{ __m }->res->content )
		);
		return $l
		
        }
        else {
                $self->{ __v }->__set_error(
                        error_message   => "unable to retrieve services: ". $self->{ __v }->{ __m }->response->message,
                        is_fatal        => 1
                )		
        }

}

1;
__END__

=head2 NAME 

WWW::VERNet::Service - Utility class for representing VERNet provisioned services.

=head2 METHODS

=head3 id ()

Returns the VERNet service ID.

=head3 description ()

Returns the VERNet service description.

=head3 type ()

Returns the service type - e.g. "Managed 10G".

=head3 has_ps ()

Returns a boolean value indicating if a provisioning statement is available for this service.

=head3 ps_link ()

Returns a the URL of the provisioning statement if a provisioning statement is available.

=head3 has_pop_data ()

Returns a boolean value indicating if live PoP data is available for this service.

=head live_data ()

Returns a L<WWW::VERNet::Service::LiveData> object exposing access to the live data available
from the VERNet PoP if PoP data is available.

=head2 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head2 SEE ALSO

L<WWW::VERNet>
L<WWW::VERNet::Service::LiveData>
L<WWW::VERNet::Service::LiveData::EndPoint>


lib/WWW/VERNet/Service/LiveData.pm
package WWW::VERNet::Service::LiveData;

use strict;
use warnings;

use DateTime;
use WWW::VERNet::Service::LiveData::EndPoint;

our @ATTR = qw( datestamp ends );
our %d = ( sun => 0, mon => 1, tue => 2, wed => 3, thu => 4, fri => 5, sat => 6 );
our %m = ( jan => 0, feb => 1, mar => 2, apr => 3, may => 4, jun => 5, jul => 6,
		aug => 7, sep => 8, oct => 9, nov => 10, dec => 12 );

sub new {
	my ( $class, $s ) = @_;

	my $self = bless {}, $class;
	$self->__init( $s );

	return $self
}

sub endpoint_ids {
	my $self = shift;

	return map { $_->eid } @{ $self->{ __endpoints } }
}

sub endpoints {
	my $self = shift;

	return @{ $self->{ __endpoints } }
}

sub endpoint {
	my ( $self, $id ) = @_;

	my ( $e ) = grep { $_->eid eq $id } @{ $self->{ __endpoints } };
	
	return $e
}

sub __init {
	my ( $self, $s ) = @_;

	# Extract the timestamp
	( $self->{ timestamp_raw } = $s->look_down( _tag => 'h3' )->as_trimmed_text ) =~ s/^.*: //;

	$self->__extract_timestamp;

	$self->__extract_live_data( $s->look_down( _tag => 'fieldset' ) );
}

sub __extract_live_data {
	my ( $self, $d ) = @_;

	$self->__extract_endpoints( $d );

	$self->{ d } = $d;
}

sub __extract_endpoints {
	my ( $self, $d ) = @_;

	my @ep = split /<u>/, $d->as_HTML;
	# Throw away the "Live PoP Data" header
	shift @ep;

	map {
		push @{ $self->{ __endpoints } }, 
			WWW::VERNet::Service::LiveData::EndPoint->new( __extract_endpoint_data( $_ ) )
	} @ep;
}

sub __extract_endpoint_data {
	my $ep = shift;

	my %ep;
	# Split into "fields" and remove HTML tags
	my @f = map { s|<.+?>||g; $_ } ( split /<strong>/, $ep );

	# Endpoint end ID
	$ep{ eid } = shift @f;
	$ep{ eid } =~ s/^From //;
	$ep{ eid } =~ s/ End://;

	# Extract endpoint field names and values
	for my $f ( @f ) {
		my( $var, $val ) = split /[:?] /, $f;
		$var =~ s/ /_/g;
		$var = lc( $var );
		$ep{ $var } = $val
	}

	return %ep
}

sub __extract_timestamp {
	my $self = shift;

	# Extract the live data timestamp and use it to create a
	# DateTime object to allow for calculation, manipulation, etc.
	my ( $wd, $d, $o, $m, $y, $t, $me ) = split / /, $self->{ timestamp_raw };

	$wd = $d{ substr lc( $wd ), 0, 3 };
	( $d ) = $d =~ /(\d+)/;
	$m = $m{ substr lc( $m ), 0, 3 };
	my ( $h, $mm, $s ) = split /:/, $t;
	$h += 12 if ( $me eq 'PM' and $h > 12 );

	$self->{ timestamp } = DateTime->new(
		year	=> $y,
		month	=> $m,
		day	=> $d,
		hour	=> $h,
		minute	=> $mm,
		second	=> $s
	);
}

1;

__END__

=head2 NAME 

WWW::VERNet::Service::LiveData - Utility class for representing VERNet services PoP data.

=head2 METHODS

=head3 endpoints ()

Returns an array of L<WWW::VERNet::Service::LiveData::EndPoint> objects representing the
PoP end points.

=head3 endpoint ( $endpoint_id )

Returns a <WWW::VERNet::Service::LiveData::Endpoint> object representing the PoP end point
identified by the provided end point ID (e.g. A001).

=head2 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head2 SEE ALSO

L<WWW::VERNet>
L<WWW::VERNet::Service>
L<WWW::VERNet::Service::LiveData::EndPoint>


lib/WWW/VERNet/Service/LiveData/EndPoint.pm
package WWW::VERNet::Service::LiveData::EndPoint;

use strict;
use warnings;

our @ATTR = qw( eid receive_power pop_transmitting link_ok loopback );

{
	no strict 'refs';

	for my $attr ( @ATTR ) {
		*{ __PACKAGE__ . "::$attr" } = sub {
			my $self = shift;
			return $self->{ $attr }
		}
	}
}

sub new {
	my ( $class, %args ) = @_;
	
	my $self = bless {}, $class;
	$self->__init( %args );

	return $self
}

sub __init {
	my ( $self, %args ) = @_;

	for my $attr ( @ATTR ) {
		$self->{ $attr } = $args{ $attr }
	}
}

1;

__END__

=head2 NAME 

WWW::VERNet::Service::LiveData::EndPoint - Utility class for representing VERNet services 
PoP end points.

=head2 METHODS

=head3 eid ()

Returns the end point ID.

=head3 receive_power ()

Returns the end point reported receive power.

=head3 pop_transmitting ()

Returns the end point PoP transmitting value.

=head3 link_ok ()

Returns the end point Link OK status.

=head3 loopback ()

Returns the end point loopback status.

=head2 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head2 SEE ALSO

L<WWW::VERNet>
L<WWW::VERNet::Service>
L<WWW::VERNet::Service::LiveData>



lib/WWW/VERNet.pm
package WWW::VERNet;

use strict;
use warnings;

use HTML::TreeBuilder;
use WWW::Mechanize;
use WWW::VERNet::Service;

our $__DOMAIN = 'support.vernet.net.au';
our $VERSION  = '0.01';

sub new {
	my ( $class, %args ) = @_;

	my $self = bless {}, $class;
	$self->__init( %args );

	return $self
}

sub __init {
	my ( $self, %args ) = @_;

	$args{ username } or die "Mandatory parameter 'username' not provided in constructor.";
	$args{ password } or die "Mandatory parameter 'password' not provided in constructor.";
	$args{ timeout  } ||= 180;

	$self->{ username } = $args{ username };
	$self->{ password } = $args{ password };
	$self->{ __m      } = WWW::Mechanize->new( timeout => $args{ timeout } );
	$self->{ __t      } = HTML::TreeBuilder->new();
	$self->__login;
}

sub service {
	my ( $self, $id ) = @_;

	$self->__set_error( 'No service ID provided to service()' )
		and return 0 unless $id;

	my ( $s ) = grep { $_->id eq $id } $self->services;

	return $s
}

sub services {
	my $self = shift;

	$self->{ __have_services } or $self->__process_services;

	return @{ $self->{ __services } }
}

sub __login {
	my $self = shift;

	my $url = "https://$__DOMAIN/index.php";

	$self->{ __m }->get( $url );

	if ( $self->{ __m }->success
		and $self->{ __m }->title
		and $self->{ __m }->title =~ /VERNet Support Portal/i
	) {
		my %f = ( username => $self->{ username },
			  password => $self->{ password }
		);

		$self->{ __m }->submit_form( with_fields => \%f );

		if ( $self->{ __m }->success
			and $self->{ __m }->res->content =~ /check your profile here/ ) {
			$self->{ __have_session } = 1;
			return 1
		}
	}
	else {
		$self->__set_error(
			error_message	=> "request was unsuccessful: ". $self->{ __m }->response->message,
			is_fatal	=> 1
		)
	}
}

sub __set_error {
        my ( $self, %a ) = @_; 

        $self->{ __have_error } = 1;
        $self->{ __last_error } = $self->{ __error };
        $self->{ __error } = $a{ error_message };
        die $self->{ __error } if ( $a{ is_fatal } )
}

sub __process_services {
	my $self = shift;

	$self->__get_services();

	# Get a list of services by grouping - e.g. <strong>Managed 1Gbe Services</strong>
	my @s = $self->{ __raw }->look_down( _tag => 'strong' );
	# Each service grouping has a <ul> of services for that group
	my @u = $self->{ __raw }->look_down( _tag => 'ul' );
	# There's a ul in the header and an empty "recently updated" service group that we can throw away
	splice( @u, 0, 2 );

	for( my $c = 0; $c < @s ; $c++ ) {
		$self->__process_service_set( $s[ $c ]->as_trimmed_text, $u[ $c ] );
	}
	
	$self->{ __have_services } = 1
}

sub __process_service_set {
	my ( $self, $service, $list ) = @_;

	for my $s ( $list->look_down( _tag => 'li' ) ) {
		push @{ $self->{ __services } }, WWW::VERNet::Service->new( $service, $s, $self )
	}
}

sub __get_services {
	my $self = shift;

	$self->{ __have_session } || $self->__login;

	my $url = "https://$__DOMAIN/myservices.php";
	$self->{ __m }->get( $url );

	if ( $self->{ __m }->success
		and $self->{ __m }->title
		and $self->{ __m }->title =~ /My Services/i
	) {
		$self->{ __raw } = $self->{ __t }->parse_content( $self->{ __m }->res->content );
	}
	else {
		$self->__set_error(
			error_message	=> "unable to retrieve services: ". $self->{ __m }->response->message,
			is_fatal	=> 1
		)
	}
}

1;

__END__

=head2 NAME

WWW::VERNet - Screen-scraping API for VERNet service information.

=head2 SYNOPSIS

This module provides a rudimentary screen-scraping "API" for VERNet service information.

You may use this module to access information about your VERNet services including
live data from end points.

	use WWW::VERNet;

        # Create a new WWW::VERNet object using our customer credentials.

        my $v = WWW::VERNet->new(
                        username => 'email@myorganisation.org',
                        password => 'My%3cretP@ssw0rd!'
                ) or die "Unable to create new WWW::VERNet object: $!\n";

        # For each VERNet service, print out the service ID and description,
        # and for service endpoint, print out the endpoint ID, link and transmit 
        # status, and receive power _if_ available.

        for my $service ( $v->services ) { 
                if ( $service->has_pop_data ) { 
                        print "Service ". $service->id ." (". $service->description .")\n";
                            
                        for my $endpoint ( $service->live_data->endpoints ) { 

                                print " - Endpoint ". $endpoint->eid 
                                        ."\n\tLink OK: ".       $endpoint->link_ok
                                        ."\n\tReceive power: ". $endpoint->receive_power
                                        ."\n\tTransmitting: ".  $endpoint->pop_transmitting
                                        ."\n\n"
                        }   
                            
                        print "\n\n"
                }   
        }

	# e.g. print something like:
	#
	# Service MYORG01234567M (CBD Campus to Remote Campus)
	# - Endpoint A001
	#	Link OK: Yes
	#	Receive power: -5.67 dBm
	#	Transmitting: -5.67 dBm
	#
	# - Endpoint A002
	#	Link OK: Yes
	#	Receive power: -8.096 dBm
	#	Transmitting: -8.096 dBm
	#
	# ...


        # If we know our service and endpoint IDs, we can get direct access
        # to the live data we are interested in.

        print "MYORG01234567M endpoint A001 Rx power is "
                . $v->service( 'MYORG01234567M' )->live_data->endpoint( 'A001' )->receive_power;

	create_service_alert( 'VERNet service MYORG01234567M link state degraded' )
		unless ( $v->service( 'MYORG01234567M' )->live_data->endpoint( 'A001' )->link_ok );

=head2 METHODS

=head3 new ( %args )

Constructor - accepts the following parameters:

=over 4

=item username

Mandatory - the username of the VERNet customer account.

=item password

Mandatory - the password of the VERNet customer account.

=item timeout

Optional - the timeout period (in seconds) to wait for HTTP requests to be service.
If not provided this will default to a value of 180s.

=back

=head3 services ()

Returns an array of L<WWW::VERNet::Service> objects representing the provisioned 
VERNet services for your account.

=head3 service ( $service_id )

Returns a L<WWW::VERNet::Service> object representing the service identified by the
B<$service_id> parameter.

=head2 AUTHOR

Luke Poskitt, C<< <ltp at cpan.org> >>

=head2 BUGS

Please report any bugs or feature requests to C<bug-www-vernet at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-VERNet>.  I will be 
notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head2 SEE ALSO

L<WWW::VERNet::Service>
L<WWW::VERNet::Service::LiveData>
L<WWW::VERNet::Service::LiveData::EndPoint>

=head2 SUPPORT

B<Please note> that this is B<not> an authorised or official API and that it is B<not>
endorsed or supported by VERNet.  This module is a convenience provided by community
support only.

You can find documentation for this module with the perldoc command.

    perldoc WWW::VERNet

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-VERNet>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-VERNet>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-VERNet>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-VERNet/>

=back

=head2 ACKNOWLEDGEMENTS

=head2 LICENSE AND COPYRIGHT

Copyright 2017 Luke Poskitt.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut
